File: Neetcode/3Sum.py
from collections import Counter
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        count = Counter(nums)
        ans = set()
        for i,num1 in enumerate(nums):
            for j,num2 in enumerate(nums[i+1:]):
                req = 0 - (num1 + num2)
                if req in count:
                    if req == num1 or req == num2:
                        if req == 0:
                            if count[req] > 2:
                                ans.add(tuple(sorted([num1, num2, req])))
                        else:
                            if count[req] > 1:
                                ans.add(tuple(sorted([num1, num2, req])))
                    else:
                        ans.add(tuple(sorted([num1, num2, req])))
        return list(ans)

File: Neetcode/3Sum.cpp
#include <algorithm>
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int l, r, sm;
        vector<vector<int>> ans;
        for (int i=0; i<nums.size(); i++){
            if (i > 0 && nums[i] == nums[i-1]){
                continue;
            }

            l = i + 1;
            r = nums.size() - 1;

            while (l < r){
                sm = nums[i] + nums[l] + nums[r];
                if (sm > 0){
                    r -= 1;
                }
                else if (sm < 0){
                    l += 1;
                }
                else{
                    ans.push_back({nums[i], nums[l], nums[r]});
                    l += 1;
                    while (nums[l] == nums[l-1] && l < r){
                        l += 1;
                    }
                }
            }
        }
        return ans;
    }
};
